crates/provider/src/provider/caller.rs

```diff
@@ -1,46 +1,42 @@
+use super::EthCallParams;
 use crate::ProviderCall;
-use alloy_json_rpc::{RpcParam, RpcReturn};
+use alloy_json_rpc::RpcReturn;
+use alloy_network::Network;
 use alloy_rpc_client::WeakClient;
-use alloy_transport::{RpcError, Transport, TransportErrorKind, TransportResult};
+use alloy_transport::{Transport, TransportErrorKind, TransportResult};
 use std::borrow::Cow;
 
-// TODO: Make `EthCall` specific. Ref: https://github.com/alloy-rs/alloy/pull/788#discussion_r1748862509.
-
 /// Trait that helpes convert `EthCall` into a `ProviderCall`.
-pub trait Caller<T, Params, Resp>: Send + Sync
+pub trait Caller<T, N, Resp>: Send + Sync
 where
     T: Transport + Clone,
-    Params: RpcParam,
+    N: Network,
     Resp: RpcReturn,
 {
     /// Method that needs to be implemented to convert to a `ProviderCall`.
     ///
-    /// This method handles serialization of the params and sends the request to relevant data
-    /// source and returns a `ProviderCall`.
+    /// This method sends the request to relevant data source and returns a `ProviderCall`.
     fn call(
         &self,
         method: Cow<'static, str>,
-        params: Params,
-    ) -> TransportResult<ProviderCall<T, serde_json::Value, Resp>>;
+        params: EthCallParams<'_, N>,
+    ) -> TransportResult<ProviderCall<T, EthCallParams<'static, N>, Resp>>;
 }
 
-impl<T, Params, Resp> Caller<T, Params, Resp> for WeakClient<T>
+impl<T, N, Resp> Caller<T, N, Resp> for WeakClient<T>
 where
     T: Transport + Clone,
-    Params: RpcParam,
+    N: Network,
     Resp: RpcReturn,
 {
     fn call(
         &self,
         method: Cow<'static, str>,
-        params: Params,
-    ) -> TransportResult<ProviderCall<T, serde_json::Value, Resp>> {
+        params: EthCallParams<'_, N>,
+    ) -> TransportResult<ProviderCall<T, EthCallParams<'static, N>, Resp>> {
```

Comment:
would be nice to have either enum for method or just separate `call`/`estimate_gas` methods so that custom implementations doesn't have to match on strings